### PR TITLE
RDRP-921 output backdata for imputation

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

The code now outputs data to used for forward imputation in the next year.

As part of this, there's a temp script to deal with the special case of 2021 data. This creates a new version of the 2021 data which has everything needed for imputation going forwards. When we've settled on how we do forward imputation, and everything is also set up in DAP, this can be deleted.

Other minor improvements to the pipeline have been included, and the precommit yaml updated so tests are excluded from limiting line lengths.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #921

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outputs/outputs_main.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 
 
 def run_outputs(  # noqa: C901
-    estimated_df: pd.DataFrame,
```
Comment (JenCheshire) [inline src/outputs/outputs_main.py]: Why isn't this needed?

File: src/site_apportionment/site_apportionment_main.py [hunk 2]
```diff
@@ -25,14 +24,12 @@ def run_site_apportionment(
     instance 0 to all other instances.
     Same percentages are used for each product group.
 
-    When running on the local network,
```
Comment (JenCheshire) [inline src/site_apportionment/site_apportionment_main.py:28]: I know this isn't part of this ticket - should we record the tidying up jobs that need doing through the pipeline?  The docstring needs updating for this function.

**Review by JenCheshire — APPROVED** [review]
Only two minor things - looks really good, really tidy.  As discussed I didn't look at the temp 2021 script

File: src/outputs/outputs_main.py (hunk 1 above)
Comment (AnneONS) [reply src/outputs/outputs_main.py]: See comment below for context, but glad you noticed, maybe I should have left this in, I'll check.

File: src/site_apportionment/site_apportionment_main.py (hunk 2 above)
Comment (AnneONS) [reply src/site_apportionment/site_apportionment_main.py:28]: Originally I was intending to do the imputation_backdata output at the end of estimation, so I wanted to do the clearing up we were going to do where we only did output of one data frame at the end of estimation. But I backtracked, and there are still some left over changes. In this case, I removed an incomplete comment.

Comment (JenCheshire) [reply src/site_apportionment/site_apportionment_main.py:28]: Great, so it's just the Args description that needs updating

**Pull request merged**